### PR TITLE
AJ-1362 - Add retries in WDS end to end test

### PR DESCRIPTION
**Pull request by yuliadub**
After reviewing all the failed runs from the past month, the following workflows cause the most failures: 
- bee creation
- workspace creation
- polling for Leo Urls (occasional 503)

Bee creation is handled by a separate workflow, so to have retries for that, changes were made in the terra wds repo. More info for that can be found here: https://github.com/DataBiosphere/terra-workspace-data-service/pull/405

For retrying workspace creation and Leo url polling, I added a simple while loop to do the retries (all other options seemed more complicated than they needed to be). Everyone using the create workspace and polling for app status functions will get these retries for free.

### CODE REVIEW

File: .github/workflows/run-e2e-tests.yaml [hunk 1]
```diff
@@ -33,7 +33,7 @@ jobs:
       - name: "Checkout dsp-reusable-workflows"
         uses: actions/checkout@v3
         with:
-          ref: main
+          ref: aj-1362-retries
```
Comment (yuliadub) [inline .github/workflows/run-e2e-tests.yaml:36]: edited just for testing

File: e2e-test/app_helper.py [hunk 2]
```diff
@@ -35,32 +35,41 @@ def poll_for_app_url(workspaceId, app_type, proxy_url_name, azure_token, leo_url
 
     # prevent infinite loop
     poll_count = 20 # 30s x 20 = 10 min
```
Comment (jladieu) [inline e2e-test/app_helper.py:37]: i know this is the existing variable name, but maybe rename to `polling_attempts_remaining`?

File: e2e-test/app_helper.py [hunk 3]
```diff
@@ -35,32 +35,41 @@ def poll_for_app_url(workspaceId, app_type, proxy_url_name, azure_token, leo_url
...
+    
+    for i in range(0,3):
+        while True:
```
Comment (jladieu) [inline e2e-test/app_helper.py:40]: Why do you need both the `while True` _and_ the for loop?

File: e2e-test/workspace_helper.py [hunk 4]
```diff
@@ -7,49 +7,59 @@
...
+                logging.info(f"Successfully started workspace creation for '{workspace_name}' in billing project '{billing_project_name}'. Workspace ID returned: {workspace_id}")
+
+                return workspace_id, data['name']
+
+            except Exception as e:
+                logging.info(f"ERROR workspace creation for '{workspace_name}' in billing project '{billing_project_name}'. Error: {e}")
+                continue
+            break
+
+    return null
```
Comment (jladieu) [inline e2e-test/workspace_helper.py:62]: What does the caller do with the null result?

File: e2e-test/app_helper.py [hunk 5]
```diff
@@ -35,32 +35,41 @@ def poll_for_app_url(workspaceId, app_type, proxy_url_name, azure_token, leo_url
...
+                                logging.error(f"{app_type} is in ERROR state. Error details: {entries['errors']}")
+                                return ""
+                            elif entries['proxyUrls'][proxy_url_name] is None:
+                                logging.error(f"{app_type} proxyUrls not found: {entries}")
+                                return ""
+                            else:
+                                logging.info(f"{app_type} is in READY state")
+                                return entries['proxyUrls'][proxy_url_name]
+                    poll_count -= 1
+                except Exception as e:
```
Comment (ashanhol) [inline e2e-test/app_helper.py:67]: (possible python ignorance) It looks like this except statement is not lined up with the try, will that be a problem?

File: e2e-test/app_helper.py [hunk 6]
```diff
@@ -35,32 +35,41 @@ def poll_for_app_url(workspaceId, app_type, proxy_url_name, azure_token, leo_url
...
+                    logging.info(f"ERROR polling for app '{app_type}' in workspace '{workspaceId}'. Error: {e}")
+                    # for retries, shorten the time for polling, since these would be caused by transient errors, not waiting for apps to start
+                    poll_count = 2
+                    continue
+                break
```
Comment (ashanhol) [inline e2e-test/app_helper.py:72]: Trying to make sure I understand, the except resets the poll count/continues the poll loop, if we exhaust our poll tries, we break out of the "while true"... but then go back into the "while true" 3 times?

File: e2e-test/workspace_helper.py [hunk 7]
```diff
@@ -7,49 +7,59 @@
...
-    # }
-    workspace_response_json = workspace_response.json()
-    data = json.loads(json.dumps(workspace_response_json))
-
-    workspace_id = data['workspaceId']
-    logging.info(f"Successfully started workspace creation for '{workspace_name}' in billing project '{billing_project_name}'. Workspace ID returned: {workspace_id}")
-
-    return workspace_id, data['name']
-
+def create_workspace(billing_project_name, azure_token, rawls_url, workspace_name = "", retries = 1):
```
Comment (ashanhol) [inline e2e-test/workspace_helper.py:10]: where is retries used?

File: e2e-test/app_helper.py (hunk 3 above)
Comment (ashanhol) [reply e2e-test/app_helper.py:40]: +1

Comment (yuliadub) [reply e2e-test/app_helper.py]: ha I thought I had a great reason to use it but when I wrote it down I realized actually it is not needed 😆

File: e2e-test/workspace_helper.py (hunk 4 above)
Comment (yuliadub) [reply e2e-test/workspace_helper.py]: the wrong thing, that shouldn't be there! good catch

File: e2e-test/app_helper.py (hunk 5 above)
Comment (yuliadub) [reply e2e-test/app_helper.py]: no, this is a good catch, I had to do a merge after I did testing and I thought I got it right but failed here. should be fixed now.

File: e2e-test/app_helper.py (hunk 6 above)
Comment (yuliadub) [reply e2e-test/app_helper.py]: as per other comment, the while true is not needed here. Added some comments to the changed logic that hopefully help understand what I am trying to do :)

File: e2e-test/workspace_helper.py (hunk 7 above)
Comment (yuliadub) [reply e2e-test/workspace_helper.py]: this was my attempt to plumb in retries, but I forgot about it midway. Added a different way to not have a hardcoded retry number everywhere

**Review by ashanhol — APPROVED** [review]
Looks good with the changes :)

File: e2e-test/workspace_helper.py [hunk 8]
```diff
@@ -4,62 +4,62 @@
...
+            workspace_id = data['workspaceId']
+            logging.info(f"Successfully started workspace creation for '{workspace_name}' in billing project '{billing_project_name}'. Workspace ID returned: {workspace_id}")
+
+            return workspace_id, data['name']
+
+        except Exception as e:
+            logging.info(f"ERROR workspace creation for '{workspace_name}' in billing project '{billing_project_name}'. Error: {e}")
+            continue
+    else:
+        # this will execute if the loop didnt exit abnormally (i.e. with break)
```
Comment (jladieu) [inline e2e-test/workspace_helper.py:61]: I got kinda confused reading this comment because it uses double-negative (didn't + abnormal), can you rephrase as a slightly more positive sentence (eg: this will execute if the loop exhausts all its retries without finding a workspace)?

File: e2e-test/workspace_helper.py [hunk 9]
```diff
@@ -4,62 +4,62 @@
...
+
+            workspace_id = data['workspaceId']
+            logging.info(f"Successfully started workspace creation for '{workspace_name}' in billing project '{billing_project_name}'. Workspace ID returned: {workspace_id}")
+
+            return workspace_id, data['name']
+
+        except Exception as e:
+            logging.info(f"ERROR workspace creation for '{workspace_name}' in billing project '{billing_project_name}'. Error: {e}")
+            continue
+    else:
```
Comment (jladieu) [inline e2e-test/workspace_helper.py:60]: This loop logic is pretty far away from where the loop begins.  Could you factor out the contents of the try block into a method so that the for/try/except/else are all within a few lines of each other?

File: e2e-test/workspace_helper.py (hunk 8 above)
Comment (jladieu) [reply e2e-test/workspace_helper.py:61]: Actually...TBH, the error message you raise does a great job explaining, so maybe you don't even need the comment?

**Comment by jladieu** [comment]
Couple suggestions but LGTM!

**Review by jladieu — APPROVED** [review]

File: e2e-test/workspace_helper.py (hunk 9 above)
Comment (yuliadub) [reply e2e-test/workspace_helper.py:60]: I struggled with this a bit - I could create a different function that is called "create_workspace_with_retries" and then move the try/catch in there and keep the create_workspace as is (since it is used by every other test in the e2e test repo) but that seemed a bit impractical.  moving code into yet another "create_workspace_smth" function feels like it could be confusing. Mainly the comment is what is making the text so long - so I considered moving it into the header of the function, but then it is less clear why it is even there. 

with all that, there is a ticket to do general test clean up, so I think this work fits well into that - so that general clean up can happen all at once and optimizations like that can be done in a consistent way.
